New regional Panantarctic example

francispoulin opened the pull request:
Following up on #106, this is a first attempt to create a regional model with ECCO-derived restoring at the boundaries.  We decided to try focusing on the ACC in the southern ocean.

It does not run yet, but after it does, it would be good to know if people agree this is a good example to include.  If yes, then we need to turn this into an example.

Commit:
creating new example for the ACC

francispoulin commented:
The error that I get is copied below.  

@simone-silvestri @glwagner 

```
[ Info: In-painting ecco salinity
[ Info: In-painting ecco salinity
ERROR: a bounds error was thrown during kernel execution on thread (65, 1, 1) in block (3, 1, 1).
Stacktrace not available, run Julia on debug level 2 for more details (by passing -g2 to the executable).

ERROR: LoadError: KernelException: exception thrown during kernel execution on device NVIDIA A100-SXM4-40GB
Stacktrace:
  [1] check_exceptions()
    @ CUDA ~/.julia/packages/CUDA/Tl08O/src/compiler/exceptions.jl:39
  [2] device_synchronize(; blocking::Bool, spin::Bool)
```

simone-silvestri commented:
what if you use the CPU and start julia with `--check-bounds=yes`?

francispoulin commented:
Thanks @simone-silvestri for the suggestion.  Will try it now.

francispoulin committed:
Update acc_regional_simulation.jl

francispoulin commented:
Things go a lot further but there is a problem with not knowing `coupled_model`.  This needs to be imported?

```
ERROR: UndefVarError: `coupled_model` not defined
Stacktrace:
 [1] top-level scope
   @ REPL[8]:1
```

francispoulin commented:
@simone-silvestri , any advice on what is going wrong here?

simone-silvestri commented:
It looks like there is a bug in the `TabulatedAlbedo` function. i.e., there is no check to make sure that we stay within bounds when interpolating in the table.
I will open a PR to fix this issue. In the meantime, if you want to proceed with the implementation without incurring in this problem, you can use `radiation = Radiation(ocean_albedo = LatitudeDependentAlbedo())`

francispoulin commented:
Thanks @simone-silvestri , I will give that a try!

francispoulin committed:
Update acc_regional_simulation.jl

Make changes so that it runs

francispoulin commented:
@simone-silvestri : I tried it and it seems like a function is not defined.

I added this at the beginning and now it seems to be running!

```
using ClimaOcean.OceanSeaIceModels.CrossRealmFluxes: LatitudeDependentAlbedo
```

simone-silvestri commented:
Ah nice. I think we can export that type.

francispoulin commented:
It's running on a CPU (i.e. slow) and still on the initial time step.

I made all these changes on the branch and can revert back to what we had previously as other fixes come along.  

Maybe I'll have something to share tomorrow.

francispoulin commented:
I started the job yesterday and it hasn't updated the output files in over 24 hours.  I think something has gone wrong.  Below is the currently display that I have.  It hasn't stopped and still running on a CPU.  Maybe we need to try it on a GPU or have more output to see what has gone wrong? Any suggestions?

```
julia> include("acc_regional_simulation.jl")
[ Info: Regridding bathymetry from existing file /u/fpoulin/.julia/scratchspaces/0376089a-ecfe-4b0e-a64f-9c555d74d754/Bathymetry/ETOPO_2022_v1_60s_N90W180_surface.nc.
┌ Warning: The westernmost meridian of `target_grid` 0.0 does not coincide with the closest meridian of the bathymetry grid, -1.4210854715202004e-14.
└ @ ClimaOcean.Bathymetry ~/software/ClimaOcean.jl/src/Bathymetry.jl:147
[ Info: In-painting ecco temperature
[ Info: In-painting ecco temperature
[ Info: In-painting ecco salinity
[ Info: In-painting ecco salinity
┌ Warning: This simulation will run forever as stop iteration = stop time = wall time limit = Inf.
└ @ Oceananigans.Simulations ~/.julia/packages/Oceananigans/dvdXO/src/Simulations/simulation.jl:55
[ Info: In-painting ecco temperature
[ Info: In-painting ecco salinity
[ Info: Initializing simulation...
[ Info: Time: 0 seconds, Iteration 0, Δt 5 minutes, max(vel): (0.00e+00, 0.00e+00, 0.00e+00), max(T): 29.70, min(T): -1.94, wtime: 2.679 minutes 
[ Info:     ... simulation initialization complete (14.029 seconds)
[ Info: Executing initial time step...
┌ Warning: Simulation stopped during initialization.
└ @ Oceananigans.Simulations ~/.julia/packages/Oceananigans/dvdXO/src/Simulations/run.jl:129

```

francispoulin commented:
My correction.  It is still running on one CPU. It is at 4 days after 7 days of computing.  Not a great ratio.

What needs to be done so we need to do to run this on a GPU?  @simone-silvestri

francispoulin commented:
> Wow, that seems quite slow! What if you move it on the GPU?

Sorry @simone-silvestri for the late reply.

I am happy to try it again on a GPU but last time there was an error.  I can try it again and let you know what the error is.

francispoulin commented:
@simone-silvestri 
I ran it on a GPU and found the following error.  It suggests I try passing -g2 when I run Julia.  I can try that but I believe I tried this before and didn't see much more.

```
ERROR: a bounds error was thrown during kernel execution on thread (225, 1, 1) in block (7, 1, 1).
Stacktrace not available, run Julia on debug level 2 for more details (by passing -g2 to the executable).
```

francispoulin commented:
I stand corrected, there is more information.  To me this actually looks very different even.

```
julia> include("acc_regional_simulation.jl")
Precompiling Oceananigans
  166 dependencies successfully precompiled in 109 seconds
Precompiling ClimaOcean
        Info Given ClimaOcean was explicitly requested, output will be shown live 
WARNING: using Units.day in module ECCO conflicts with an existing identifier.
  204 dependencies successfully precompiled in 143 seconds. 168 already precompiled.
  2 dependencies had output during precompilation:
┌ ClimaOcean
│  [Output was shown above]
└  
┌ Accessors → AccessorsUnitfulExt
│  [pid 760940] waiting for IO to finish:
│   Handle type        uv_handle_t->data
│   fs_event           0x25d5fe0->0x7fdd5f8fbeb0
│   timer              0x2385440->0x7fdd5f8fbee0
│  This means that a package has started a background task or event source that has not finished running. For precompilation to complete successfully, the event source needs to be closed explicitly. See the developer documentation on fixing precompilation hangs for more help.
└  
[ Info: Regridding bathymetry from existing file /u/fpoulin/.julia/scratchspaces/0376089a-ecfe-4b0e-a64f-9c555d74d754/Bathymetry/ETOPO_2022_v1_60s_N90W180_surface.nc.
┌ Warning: The westernmost meridian of `target_grid` 0.0 does not coincide with the closest meridian of the bathymetry grid, -1.4210854715202004e-14.
└ @ ClimaOcean.Bathymetry ~/software/ClimaOcean.jl/src/Bathymetry.jl:147
[ Info: In-painting ecco temperature
[ Info: In-painting ecco temperature
[ Info: In-painting ecco salinity
[ Info: In-painting ecco salinity
ERROR: a bounds error was thrown during kernel execution on thread (1, 1, 1) in block (3, 1, 1).
Stacktrace:
 [1] indexed_iterate at ./tuple.jl:92
 [2] indexed_iterate at ./tuple.jl:92
 [3] stateindex at /u/fpoulin/software/ClimaOcean.jl/src/ClimaOcean.jl:40
 [4] ECCORestoring at /u/fpoulin/software/ClimaOcean.jl/src/DataWrangling/ecco_restoring.jl:210
 [5] DiscreteForcing at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Forcings/discrete_forcing.jl:51
 [6] hydrostatic_free_surface_tracer_tendency at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:133
 [7] macro expansion at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl:240
 [8] gpu_compute_hydrostatic_free_surface_Gc! at /u/fpoulin/.julia/packages/KernelAbstractions/QE5mt/src/macros.jl:95
 [9] gpu_compute_hydrostatic_free_surface_Gc! at ./none:0

ERROR: a bounds error was thrown during kernel execution on thread (1, 1, 1) in block (67, 1, 1).
Stacktrace:
 [1] indexed_iterate at ./tuple.jl:92
 [2] indexed_iterate at ./tuple.jl:92
 [3] stateindex at /u/fpoulin/software/ClimaOcean.jl/src/ClimaOcean.jl:40
 [4] ECCORestoring at /u/fpoulin/software/ClimaOcean.jl/src/DataWrangling/ecco_restoring.jl:210
 [5] DiscreteForcing at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Forcings/discrete_forcing.jl:51
 [6] hydrostatic_free_surface_tracer_tendency at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:133
 [7] macro expansion at /u/fpoulin/.julia/packages/Oceananigans/dvdXO/src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl:240
Unhandled Task ERROR: KernelException: exception thrown during kernel execution on device NVIDIA A100-SXM4-40GB
```

francispoulin commented:
> Can you make an MWE for this and open an issue?

I will certainly give it a try and see what part of it is causing the issue.  This will likely take me a day or two to get to.

francispoulin commented:
@simone-silvestri , I realize it's been a few months but I am still keen to this this example up and running. 

I can try this all again this week but if you had time to meet for an hour, I wonder if that would help?

simone-silvestri commented:
Sure, I ll text on slack.

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

Commit:
updated script

simone-silvestri committed:
add different masking

Commit:
updating script to get further

Commit:
fix node problem

francispoulin committed:
fix phinodes

francispoulin committed:
add grid to varphinodes

Commit:
fix varphinode for real

Commit:
a working version on a gpu

francispoulin commented:
I'm happy to say that @simone-silvestri and I got this example working on a GPU.  

I ran it on a coarse grid and it looks reasonable. I'm now running it on a more reasonable grid.  I hope to have some resutls to share with everyone tomorrow.

If anyone has any suggestions about the code please let me know.  @glwagner.  We currently include -80 to -20 degrees but it doesn't have to be that.

francispoulin commented:
The good news is that things are running on a GPU.  Below are the results for the first 30 days.

The bad news is that things are very slow.  For comparison, the near global ocean model does the siulation of 1 day in 16 minutes.  In contrast, this coupled regional model does 1 day in 78 minutes.  That is more than 4 times slower.  

I will keep this running as long as I can but can someone tell me how much slower we should expect the slower model to be?  I saw that it used a time step of 10 seconds, but didn't have any output for the global uncoupled model and don't know what time step it took.

@simone-silvestri @glwagner 

https://github.com/user-attachments/assets/7e0e4593-7336-42df-a8c2-2e22c742710d

https://github.com/user-attachments/assets/a7ba927c-97d8-48bb-a961-34fcc6a89ad7

https://github.com/user-attachments/assets/6be49bf6-9b32-41fb-87b9-0442d35da13d

glwagner commented:
What GPU?

francispoulin commented:
I'm running it on A100's, so it should be fast. @simone-silvestri suggested I change the time step to 10 minutes instead of 10 seconds.  Working on that now. That should solve the problem.

glwagner commented:
Haha yeah, that will give you a factor of 60x right away!

Commit:
change time step to 10 minutes

Commit:
cleaning up callbacks

francispoulin commented:
Changing the time step made a world of difference.  I used 1 minute for the first 10 days and then 10 minutes for the remaining two years.  Here are the results for the temperature and turbulent kinetic energy.  I can't post the speed as it's larger than 100 MB.  

One observation is there is a lot of unbalanced motions.  But on the bright side we do see eddies forming and interesting stuff is happening.

I presume the next step is to spin this up.  If yes, for long how?  I have heard people mention 100 years but I'm not sure what is standard.

Any thoughts are welcome!

https://github.com/user-attachments/assets/5715eb61-3aaa-4516-a7ce-922a2b510720

https://github.com/user-attachments/assets/f3df0771-a66c-4150-9c33-aa6a10c361aa

simone-silvestri commented:
Very nice. I am wondering if the north restoring is a bit too strong. 
The change in density is quite visible and it seems to spin up baroclinic eddies due to instabilities at the edge of the sponge region during winter periods. Maybe we should reduce the sponge region? Or use a higher-order mask rather than a linear one?

simone-silvestri commented on examples/acc_regional_simulation.jl:
```diff
@@ -0,0 +1,219 @@
+ 
+grid = ImmersedBoundaryGrid(grid, GridFittedBottom(bottom_height), active_cells_map=true) 
+
+dates = DateTimeProlepticGregorian(1993, 1, 1) : Month(1) : DateTimeProlepticGregorian(1993, 5, 1)
```
We are using only 5 months for restoring here.
Therefore, the restoring and the forcing go out of sync passed 5 months and we create strong density gradients in the northern sponge region. Most likely we should include a whole year.
```suggestion
dates = DateTimeProlepticGregorian(1993, 1, 1) : Month(1) : DateTimeProlepticGregorian(1993, 12, 1)
```

francispoulin committed:
Update examples/acc_regional_simulation.jl

Co-authored-by: Simone Silvestri <silvestri.simone0@gmail.com>

francispoulin commented:
Thanks @simone-silvestri .  I ran this for 10 years and it seems to wokr fine, but the animations are getting rather large.  

Do you want me to make an animation outputting every week or so?

Or I could try running it for 100 years and output it only a few times a week.  

Let me know what you would prefer to see.

simone-silvestri commented:
Is it too large a video of surface quantity every 2 days maybe? A week might be too much probably

francispoulin commented:
This is a plot of the temperature, outputting every day.  It's the smallest of the three files, only 48 MB.  I can share more if you like.

https://github.com/user-attachments/assets/e5559fc7-dd9f-476e-8c6b-2434269c78f0

Do you think the sponge is any better here?

simone-silvestri commented:
A little better but it does not seem good enough. I can easily see the demarcation line where the sponge is active. We probably should either reduce the restoring timescale or the size of the sponge region that is quite large

francispoulin commented:
Thanks @simone-silvestri .  Did you want to change the script however you think is best and then I can run it?  We should probably output less often if we want to run it for 10 years.

simone-silvestri commented:
Sure, I can reduce the restoring rate and sponge region and output once every 5 days

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

simone-silvestri committed:
change to the restoring rate and the outputwriter schedule

simone-silvestri commented:
I have reduced a bit the restoring rate, reduced the extend of the northern sponge region, and changed the output writer schedule to 5 days

francispoulin commented:
Thanks @simone-silvestri .  I just updated my repo and will run it right away.  I hope to have some results by tomorrow, maybe.

francispoulin commented:
When I tried running it I got an error.  Maybe this is related to other updates?

```
ERROR: LoadError: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore should be avoided.

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
```

navidcy commented:
> When I tried running it I got an error. Maybe this is related to other updates?
> 
> ```
> ERROR: LoadError: Scalar indexing is disallowed.
> Invocation of getindex resulted in scalar indexing of a GPU array.
> This is typically caused by calling an iterating implementation of a method.
> Such implementations *do not* execute on the GPU, but very slowly on the CPU,
> and therefore should be avoided.
> 
> If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
> to enable scalar iteration globally or for the operations in question.
> ```

Yeah, see https://github.com/CliMA/ClimaOcean.jl/pull/318#issuecomment-2585500310, https://github.com/CliMA/Oceananigans.jl/pull/4037, https://github.com/CliMA/Oceananigans.jl/issues/4036

We need to fix it!

francispoulin commented:
Thanks @navidcy for pointing this out.  I will look to see if there is a bandaide fix in the mean time.

navidcy commented:
> Thanks @navidcy for pointing this out. I will look to see if there is a bandaide fix in the mean time.

https://github.com/CliMA/Oceananigans.jl/issues/4036#issuecomment-2588140094 as bandaid 🩹?

simone-silvestri commented:
I think we need to update Climaocean to Oceananigans version 0.95.6 to make sure the GPUArrays that comes with it is correct

francispoulin commented:
Thanks @navidcy for pointing this out.  I tried the example and can get the same error, but I don't see the fix anywhere in the comment.  Is there an easy temporary fix?

francispoulin commented:
I presume we still need another patch before we can move forward with this example?

francispoulin commented:
@simone-silvestri , can we chat about this sometime?

simone-silvestri commented:
Sure, do you want me to update the example with the current syntax in the meantime?

simone-silvestri committed:
Merge remote-tracking branch 'origin/main' into fjp/acc_regional_model

simone-silvestri committed:
update example

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

simone-silvestri committed:
run docs

simone-silvestri committed:
run docs

simone-silvestri commented on examples/acc_regional_simulation.jl:
```diff
@@ -0,0 +1,223 @@
+arch = GPU() 
+
+Nx = 1440
+Ny = 600
```
```suggestion
Ny = 400
```

simone-silvestri committed:
Update examples/acc_regional_simulation.jl

simone-silvestri committed:
add a dataset

simone-silvestri committed:
fix restoring

simone-silvestri committed:
jld2 writer

simone-silvestri committed:
remove the interior

simone-silvestri committed:
correctly compute restoring on GPU

simone-silvestri committed:
pass it

francispoulin committed:
Update acc_regional_simulation.jl

Make sure to only use ECCCO4Monthly, should remove numerical instabilities we were getting with ECCO2Daily.

francispoulin committed:
updating file names

Commit:
changing the name of the files to save

simone-silvestri committed:
try with passing the grid

Commit:
modifying some names and using the visualization from near global simulation

francispoulin committed:
changing a few names and using the visualization from near global (for real this time)

francispoulin committed:
change coupled_simulation to simulation

francispoulin committed:
remove two unnecessary lines

francispoulin committed:
fix alignments

francispoulin committed:
restore backend definition

francispoulin committed:
returning to model

francispoulin committed:
changing initial time step to 2*minutes

simone-silvestri committed:
Merge remote-tracking branch 'origin/main' into fjp/acc_regional_model

Commit:
trying the simulation from T=S=0 and no atmospheric forcing

Commit:
adding line so it can run

simone-silvestri committed:
Update Project.toml

simone-silvestri committed:
Update near_global_ocean_simulation.jl

simone-silvestri committed:
Update one_degree_simulation.jl

simone-silvestri committed:
changes

simone-silvestri committed:
force with salinity field

simone-silvestri committed:
Merge branch 'ss/adapt-to-new-oceananigans' into fjp/acc_regional_model

simone-silvestri committed:
Merge branch 'fjp/acc_regional_model' of github.com:CliMA/ClimaOcean.jl into fjp/acc_regional_model

simone-silvestri committed:
adding the Manifest

simone-silvestri committed:
add CUDA

simone-silvestri committed:
change the zfaces

simone-silvestri committed:
add CUDA to the examples

simone-silvestri committed:
using CUDA

simone-silvestri committed:
Merge branch 'fjp/acc_regional_model' of github.com:CliMA/ClimaOcean.jl into fjp/acc_regional_model

simone-silvestri commented:
The preview will be found here https://clima.github.io/ClimaOceanDocumentation/previews/PR142/

simone-silvestri committed:
revert cuda

francispoulin committed:
specify which record function to use

francispoulin committed:
fixing names of the files to load

francispoulin commented:
This is the result after 730 days.   The great news is that the regional mode is numerically stable!  Things looks reasonably good to me.  Do others have any concerns?

The KE plot is not very interesting so maybe I should fix the range of the colorbar.  

Uploading acc_region_surface.mp4…

navidcy committed:
Update examples/acc_regional_simulation.jl

navidcy commented:
Looks awesome!
What are the stripes at the northern boundary?

glwagner commented:
It's very nice! In the visualization, I don't see any action in the TKE field. It would also be nicer to use a wider aspect ratio.

navidcy committed:
Delete Manifest.toml

navidcy committed:
bump Oceananigans compat

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy commented:
Reduce more?! Let's see... Perhaps there is an outlier value somewhere and that's where the 0.002 is coming but everything else is 1e-5?

glwagner commented:
Are you plotting the bottom vertical level or somethign?

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
Update make.jl

navidcy commented:
> Are you plotting the bottom vertical level or something?

This

https://github.com/CliMA/ClimaOcean.jl/blob/ed80673ee279d9f88fdceef9c34dd953346b8a3c/examples/acc_regional_simulation.jl#L122

suggest that indeed the surface is being plotted.

Strange!

navidcy commented:
But there is no  winds here in this example. It's a free run.

So that's why CATKE is not being giving high diffusivities?

francispoulin commented:
I agee this is very strange @navidcy . 

We are forcing the temperature, salinity and horizontal velocities.

navidcy commented:
> I agee this is very strange @navidcy .
> 
> We are forcing the temperature, salinity and horizontal velocities. See below.
> 
> ```
> forcing = (T = DatasetRestoring(T_meta, grid; rate=1/5days, mask=tracer_mask),
> 	       S = DatasetRestoring(S_meta, grid; rate=1/5days, mask=tracer_mask),
> 	       u = Forcing(u_restoring; discrete_form=true, parameters=(; rate=1/5days)),
> 	       v = Forcing(v_restoring; discrete_form=true, parameters=(; rate=1/5days)))
> ```

Actually, when I noticed that there is no wind stress forcing (either via momentum flux  boundary condition nor via winds) then I was implying that perhaps that's where the difference is? Only saying this because it seems that you were looking at the one-degree/quarter-degree simulations (which include momentum forcing) and expect to get similar-order of magnitude of TKE in this case that it doesn't?

I'm not sure there is nothing else going wrong but wanted to point this difference in setups which might be the culprit.

Commit:
a working version with non constant TKE

Commit:
adding a regional acc example, with TKE

francispoulin committed:
removing testing folder

navidcy commented:
Cool! I can help with #595!

navidcy commented on examples/acc_regional_simulation.jl:
```diff
@@ -0,0 +1,237 @@
+
+ocean = ocean_simulation(grid; forcing, momentum_advection, tracer_advection)
+
+set!(ocean.model, T=T_meta[1], S=S_meta[1])
```
```suggestion
set!(ocean.model, T=first(T_meta), S=first(S_meta))
```

francispoulin commented on examples/acc_regional_simulation.jl:
I like the name and think this is better.  Not everyone knows what the ACC is.   I'll make the change.

francispoulin committed:
Update examples/acc_regional_simulation.jl

Co-authored-by: Navid C. Constantinou <navidcy@users.noreply.github.com>

francispoulin committed:
Changing the name of the example

francispoulin committed:
Merge branch 'main' of github.com:CliMA/ClimaOcean.jl into fjp/acc_regional_model

francispoulin committed:
Merge branch 'fjp/acc_regional_model' of github.com:CliMA/ClimaOcean.jl into fjp/acc_regional_model

navidcy commented on docs/make.jl:
```diff
@@ -17,9 +17,10 @@ const EXAMPLES_DIR = joinpath(@__DIR__, "..", "examples")
+    # "single_column_os_papa_simulation.jl",
+    # "one_degree_simulation.jl",
+    # "near_global_ocean_simulation.jl"
+    "acc_regional_simulation.jl"
```
```suggestion
    "panantarctic_regional_simulation.jl"
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,237 @@
+using Printf
+using CUDA
+
+Oceananigans.TimeSteppers.time_step!(::Nothing, dt) = nothing
```
do we need this?

francispoulin replied:
I think we need the `using CUDA` since that's how GPUs work now adays, or so I thought.

I don't think the other two are needed through.  I will try it tomorrow without the three and see if it still runs.

navidcy replied:
I only meant to comment on the

```Julia
Oceananigans.TimeSteppers.time_step!(::Nothing, dt) = nothing
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,237 @@
+Nz = 40
+
+depth = 6000meters
+z = ExponentialCoordinate(Nz, -depth)
```
```suggestion
z = ExponentialCoordinate(Nz, -depth, 0)
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,237 @@
+
+# And now we make a movie:
+
+CairoMakie.record(fig, "acc_region_surface.mp4", 1:Nt, framerate = 8) do nn
+    n[] = nn
+end
+nothing #hide
+
+# ![](near_global_ocean_surface.mp4)
```
```suggestion
CairoMakie.record(fig, "panantarctic_regional_surface.mp4", 1:Nt, framerate = 8) do nn
    n[] = nn
end
nothing #hide

# ![](panantarctic_regional_surface.mp4)
```

francispoulin committed:
Update docs/make.jl

Co-authored-by: Navid C. Constantinou <navidcy@users.noreply.github.com>

francispoulin committed:
Update docs/make.jl

Co-authored-by: Navid C. Constantinou <navidcy@users.noreply.github.com>

francispoulin committed:
Update examples/panantarctic_regional_simulation.jl

Co-authored-by: Navid C. Constantinou <navidcy@users.noreply.github.com>

francispoulin committed:
Update examples/panantarctic_regional_simulation.jl

Co-authored-by: Navid C. Constantinou <navidcy@users.noreply.github.com>

francispoulin committed:
Update panantarctic_regional_simulation.jl

francispoulin committed:
uncomment the running of the simulation

francispoulin commented:
Sorry for the long wait everyone but here is an animation of the first two years, and it looks reasoanble to me.  

Note that the mp4 that was created is over 200 Mb so I had to reduce it.

Any other concerns about this example?

https://github.com/user-attachments/assets/254db60f-c923-4197-b4a8-e756143f9155

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+# Now that the simulation has spun up, we can run it for the full 2 years.
+# We increase the maximum time step size to 10 minutes and let the simulation run for 2 years.
+
+simulation.stop_time = 2*365days
```
```suggestion
simulation.stop_time = 60days
```

navidcy replied:
I'm merging this so we can hopefully have a docs preview

navidcy commented on examples/panantarctic_regional_simulation.jl:
I think the example needs some text at the top explaining what this is about etc, right?

navidcy replied:
done

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+start_date = DateTime(1993, 1, 1)
+end_date   = DateTime(1993, 12, 1) 
+
+#
+# Restoring force 
+#               φS                   φN             -20
+# -------------- | ------------------ | ------------ |
+# no restoring   0    linear mask     1   mask = 1   1
+#
```
How would this appear when literated? We can wait for the preview and see.

navidcy replied:
https://clima.github.io/ClimaOceanDocumentation/previews/PR142/literated/panantarctic_regional_simulation/

navidcy committed:
run less time

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+
+fig, ax, hm = heatmap(h, colormap=:deep, colorrange=(-depth, 0))
+Colorbar(fig[0, 1], hm, label="Bottom height (m)", vertical=false)
+save("acc_bathymetry.png", fig)
```
```suggestion
save("panantarctic_bathymetry.png", fig)
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+# In particular, we generate an animation of the evolution of surface fields:
+# surface speed (s), surface temperature (T), and turbulent kinetic energy (e).
+
+u = FieldTimeSeries("acc_surface_fields.jld2", "u"; backend = OnDisk())
+v = FieldTimeSeries("acc_surface_fields.jld2", "v"; backend = OnDisk())
+T = FieldTimeSeries("acc_surface_fields.jld2", "T"; backend = OnDisk())
+e = FieldTimeSeries("acc_surface_fields.jld2", "e"; backend = OnDisk())
```
```suggestion
u = FieldTimeSeries("panantarctic_surface_fields.jld2", "u"; backend = OnDisk())
v = FieldTimeSeries("panantarctic_surface_fields.jld2", "v"; backend = OnDisk())
T = FieldTimeSeries("panantarctic_surface_fields.jld2", "T"; backend = OnDisk())
e = FieldTimeSeries("panantarctic_surface_fields.jld2", "e"; backend = OnDisk())
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+
+backend    = JRA55NetCDFBackend(41) 
+atmosphere = JRA55PrescribedAtmosphere(arch; backend)
+radiation  = Radiation()
```
```suggestion
radiation  = Radiation(arch)
```

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+     return - p.rate * fields.v[i, j, k] * northern_mask(φ)
+end
+
+T_meta = Metadata(:temperature; start_date, end_date, dataset = ECCO4Monthly())
+S_meta = Metadata(:salinity;    start_date, end_date, dataset = ECCO4Monthly())
```
I'm debating whether to use ECCO2 here for two reasons:

- it's 1/4 resolution so much more vigorous eddy field
- to differentiate this example from the near global
- perhaps we are able to also initialise u and v?

navidcy replied:
But we can also leave this change for a later PR!

glwagner replied:
For sponge layers that are pure relaxation, we typically want to relax to climatology rather than an eddying solution. If we have proper open boundaries (which is not supported yet), then it will make sense to pair the open boundary condition + sponge layer with the resolution of the simulation. (But this comment could only be referring to initial conditions?)

navidcy committed:
Apply suggestions from code review

navidcy commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+     wall_time[1] = time_ns()
+end
+
+simulation.callbacks[:progress] = Callback(progress, TimeInterval(6hours)) 
```
```suggestion
simulation.callbacks[:progress] = Callback(progress, TimeInterval(5days)) 
```

navidcy committed:
try ECCO2

navidcy committed:
Update examples/panantarctic_regional_simulation.jl

navidcy committed:
Update panantarctic_regional_simulation.jl

navidcy committed:
Apply suggestions from code review

navidcy committed:
fix phrasing

navidcy committed:
literate

navidcy committed:
bring back everyone in the show

navidcy committed:
there is sea ice

navidcy committed:
dreaded missing comma

francispoulin committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
Update make.jl

francispoulin commented on examples/panantarctic_regional_simulation.jl:
```diff
@@ -0,0 +1,233 @@
+forcing = (T = DatasetRestoring(T_meta, grid; rate=1/5days, mask=tracer_mask),
+	       S = DatasetRestoring(S_meta, grid; rate=1/5days, mask=tracer_mask),
+	       u = Forcing(u_restoring; discrete_form=true, parameters=(; rate=1/5days)),
+	       v = Forcing(v_restoring; discrete_form=true, parameters=(; rate=1/5days)))
```
Good questions, and thanks for spelling it out.  

@simone-silvestri was the one that came up with this approach and I know that it works but happy to try something else, if you think it's better/easier.  

What you suggests is not only shorter but more intuitive.   I'm happy for us to try it and if it works to stick with your proposed changes.

simone-silvestri replied:
For question 1, we are restoring to a zero field (basically just damping the velocities) to we cannot use `DatasetRestoring`, which is intended to restore to a specific dataset. 
Regarding 2, we could use `LinearlyTaperedPolarMask` in the forcing. We need two different masks (one for the tracer and one for the velocity), and this difference, in my opinion, is quite clear using functions.
However we could use `LinearlyTaperedPolarMask` and see if it becomes more clear

navidcy replied:
I'm covered!

navidcy committed:
Update panantarctic_regional_simulation.jl

francispoulin committed:
Update panantarctic_regional_simulation.jl

navidcy committed:
Rephrase for clarity

navidcy committed:
Update make.jl

navidcy commented:
With all the examples on the docs built fails with a memory error:

https://buildkite.com/clima/climaocean-examples/builds/4137#0198afd0-9440-43a6-8940-24c14f742694/30-1301

But the Panantarctic example runs all good when I comment out the rest of the examples... How come? Somehow memory usage from previous examples doesn't get freed up?

navidcy committed:
Update make.jl

francispoulin commented:
Thanks for being the brave one to try it @navidcy .  I seems that all checks are successful exept for one that is pending.  It seems strange that it's pending after so long.

glwagner commented:
It's pending because tartarus is down :-(

francispoulin commented:
> It's pending because tartarus is down :-(

Ah, good to know!  I hope all tests will pass after tartarus is back up.  

If/When that happens, does anyone want us to do anything else for this example?

francispoulin committed:
updating resolution for Ny

glwagner commented:
1/4 resolution is pretty good for an example!

simone-silvestri commented:
> I was only able to og as high as 1/10th of a degree but looking at this I can't see any differences with the 1/4th of a degree, at least for the first 60 days. I guess we shouldn't expect too much of a difference because of a factor of only 2.5.
> 
I think much of what we see is the initial conditions; maybe if we were to initialize from rest, there would be much more difference between the 1/4 and the 1/10th degree.

navidcy committed:
Change interpolation passes from 7 to 5

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

simone-silvestri committed:
Merge branch 'main' into fjp/acc_regional_model

glwagner commented:
I think CI needs to be re-run

francispoulin committed:
Update panantarctic_regional_simulation.jl

adding a comment

navidcy commented:
Getting similar memory errors like, e.g., 

https://buildkite.com/clima/climaocean-examples/builds/4289#01991660-b828-4efe-b4ab-6ad2ea7a519f/23-1871

in #591. Trying to use `GC.gc(); CUDA.reclaim()` after each example. See https://github.com/CliMA/ClimaOcean.jl/pull/591/commits/46930966cd19b449c87e10782e58fc19c9d80238

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

glwagner commented:
Actually, it looks like we were allowing 4 parallel builds at a time. I changed this to just 2 parallel builds.

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
Update make.jl

navidcy committed:
Update make.jl

francispoulin commented:
Thanks @navidcy for all of your help with this and I'm sorry that it's such a pain.  I am also very confused by what's going on as we are only adding an example, not changing the inner workings of the library.

Not knowing what else to do either, I started with the first error `buildkite/climateocean-ci`.  I saw what lines were exectued and I tried them.  On my server it worked fine.  Yay!   On my laptop and it failed.  Below is the output of my not so successful attempt.   It seems like this is in what some have called dependency hell.  Might we be having this problem here?  

I'm not bothered by this on my laptop, as I can just delete `.julia` and start fresh, but I'm trying to think of what's going on with these tests and why they are failing.  And it seems to change between 2 and 4 that fail. It's not even consistant.  Hmm....

```
$ echo '--- Instantiate project'
julia --project -e 'using Pkg; Pkg.instantiate(verbose=true)'
echo '--- Precompile project'
julia --project -e 'using Pkg; Pkg.precompile(strict=true)'
julia --project -e 'using Pkg; Pkg.status()'
echo '--- Initialize ClimaOcean test suite'
julia --project -e 'using CUDA; CUDA.precompile_runtime()'
julia --project -e 'using Pkg; Pkg.test()'
--- Instantiate project
┌ Warning: The project dependencies or compat requirements have changed since the manifest was last resolved.
│ It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
└ @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:1142
Precompiling packages finished.
  10 dependencies successfully precompiled in 39 seconds. 229 already precompiled.
  1 dependency had output during precompilation:
┌ ClimaOcean
│  WARNING: could not import Parameters.Rv_over_Rd into PrescribedAtmospheres
│  WARNING: could not import Parameters.Rv_over_Rd into InterfaceComputations
└  
--- Precompile project
┌ Warning: The project dependencies or compat requirements have changed since the manifest was last resolved.
│ It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
└ @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:1142
Project ClimaOcean v0.8.4
Status `~/Software/ClimaOcean.jl/Project.toml`
  [79e6a3ab] Adapt v4.3.0
  [179af706] CFTime v0.2.1
  [052768ef] CUDA v5.8.2
  [6ba0ff68] ClimaSeaIce v0.3.3
  [9c784101] CubicSplines v0.2.1
  [124859b0] DataDeps v0.7.13
  [ffbed154] DocStringExtensions v0.9.5
  [787d08f9] ImageMorphology v0.4.6
  [033835bb] JLD2 v0.5.15
  [63c18a36] KernelAbstractions v0.9.38
  [da04e1cc] MPI v0.20.23
  [85f8d34a] NCDatasets v0.14.8
  [9e8cae18] Oceananigans v0.97.5
  [6fe1bfb0] OffsetArrays v1.17.0
⌅ [aea7be01] PrecompileTools v1.2.1
  [6c6a2e73] Scratch v1.3.0
  [d496a93d] SeawaterPolynomials v0.3.10
  [90137ffa] StaticArrays v1.9.14
  [49b00bb7] SurfaceFluxes v0.12.0
⌅ [b60c26fb] Thermodynamics v0.12.15
  [a5390f91] ZipFile v0.10.1
  [ade2ca70] Dates
  [f43a241f] Downloads v1.6.0
  [de0858da] Printf
  [10745b16] Statistics v1.10.0
Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated`
Warning The project dependencies or compat requirements have changed since the manifest was last resolved. It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
--- Initialize ClimaOcean test suite
┌ Warning: The project dependencies or compat requirements have changed since the manifest was last resolved.
│ It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
└ @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:1142
     Testing ClimaOcean
┌ Warning: Could not use exact versions of packages in manifest, re-resolving
└ @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1829
ERROR: Unsatisfiable requirements detected for package ClimaSeaIce [6ba0ff68]:
 ClimaSeaIce [6ba0ff68] log:
 ├─possible versions are: 0.1.0-0.3.3 or uninstalled
 └─restricted to versions 0.3.5-0.3 by project [07d2ffc0] — no versions left
   └─project [07d2ffc0] log:
     ├─possible versions are: 0.0.0 or uninstalled
     └─project [07d2ffc0] is fixed to version 0.0.0
Stacktrace:
  [1] check_constraints(graph::Pkg.Resolve.Graph)
    @ Pkg.Resolve ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Resolve/graphtype.jl:998
  [2] Pkg.Resolve.Graph(compat::Dict{Base.UUID, Dict{VersionNumber, Dict{Base.UUID, Pkg.Versions.VersionSpec}}}, compat_weak::Dict{Base.UUID, Dict{VersionNumber, Set{Base.UUID}}}, uuid_to_name::Dict{Base.UUID, String}, reqs::Dict{Base.UUID, Pkg.Versions.VersionSpec}, fixed::Dict{Base.UUID, Pkg.Resolve.Fixed}, verbose::Bool, julia_version::VersionNumber)
    @ Pkg.Resolve ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Resolve/graphtype.jl:345
  [3] deps_graph(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, uuid_to_name::Dict{Base.UUID, String}, reqs::Dict{Base.UUID, Pkg.Versions.VersionSpec}, fixed::Dict{Base.UUID, Pkg.Resolve.Fixed}, julia_version::VersionNumber, installed_only::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:600
  [4] resolve_versions!(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, julia_version::VersionNumber, installed_only::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:413
  [5] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, level::Pkg.Types.UpgradeLevel; skip_writing_project::Bool, preserve::Nothing)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1553
  [6] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; level::Pkg.Types.UpgradeLevel, mode::Pkg.Types.PackageMode, preserve::Nothing, update_registry::Bool, skip_writing_project::Bool, kwargs::@Kwargs{io::Base.DevNull})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:351
  [7] up
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:326 [inlined]
  [8] up
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:164 [inlined]
  [9] #resolve#143
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:357 [inlined]
 [10] resolve
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:356 [inlined]
 [11] (::Pkg.Operations.var"#117#122"{String, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.PackageSpec})()
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1831
 [12] with_temp_env(fn::Pkg.Operations.var"#117#122"{String, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.PackageSpec}, temp_env::String)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1721
 [13] (::Pkg.Operations.var"#115#120"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String})(tmp::String)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1810
 [14] mktempdir(fn::Pkg.Operations.var"#115#120"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String}, parent::String; prefix::String)
    @ Base.Filesystem ./file.jl:766
 [15] mktempdir(fn::Function, parent::String)
    @ Base.Filesystem ./file.jl:762
 [16] mktempdir
    @ ./file.jl:762 [inlined]
 [17] sandbox(fn::Function, ctx::Pkg.Types.Context, target::Pkg.Types.PackageSpec, target_path::String, sandbox_path::String, sandbox_project_override::Pkg.Types.Project; preferences::Dict{String, Any}, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1768
 [18] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1971
 [19] test
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1915 [inlined]
 [20] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, test_fn::Nothing, julia_args::Cmd, test_args::Cmd, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool, kwargs::@Kwargs{io::Base.TTY})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:444
 [21] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.TTY, kwargs::@Kwargs{})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:159
 [22] test(pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:148
 [23] test(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::@Kwargs{})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:174
 [24] test()
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:165
 [25] top-level scope
    @ none:1

caused by: empty intersection between Thermodynamics@0.12.15 and project compatibility 0.14
Stacktrace:
  [1] resolve_versions!(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, julia_version::VersionNumber, installed_only::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:396
  [2] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, level::Pkg.Types.UpgradeLevel; skip_writing_project::Bool, preserve::Nothing)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1553
  [3] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; level::Pkg.Types.UpgradeLevel, mode::Pkg.Types.PackageMode, preserve::Nothing, update_registry::Bool, skip_writing_project::Bool, kwargs::@Kwargs{io::Base.DevNull})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:351
  [4] up
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:326 [inlined]
  [5] up
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:164 [inlined]
  [6] #resolve#143
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:357 [inlined]
  [7] resolve
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:356 [inlined]
  [8] (::Pkg.Operations.var"#117#122"{String, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.PackageSpec})()
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1823
  [9] with_temp_env(fn::Pkg.Operations.var"#117#122"{String, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.PackageSpec}, temp_env::String)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1721
 [10] (::Pkg.Operations.var"#115#120"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String})(tmp::String)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1810
 [11] mktempdir(fn::Pkg.Operations.var"#115#120"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#130#134"{Bool, Cmd, Cmd, Nothing, Pkg.Types.Context, Vector{Tuple{String, Base.Process}}, String, Pkg.Types.PackageSpec}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String}, parent::String; prefix::String)
    @ Base.Filesystem ./file.jl:766
 [12] mktempdir(fn::Function, parent::String)
    @ Base.Filesystem ./file.jl:762
 [13] mktempdir
    @ ./file.jl:762 [inlined]
 [14] sandbox(fn::Function, ctx::Pkg.Types.Context, target::Pkg.Types.PackageSpec, target_path::String, sandbox_path::String, sandbox_project_override::Pkg.Types.Project; preferences::Dict{String, Any}, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1768
 [15] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
    @ Pkg.Operations ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1971
 [16] test
    @ ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1915 [inlined]
 [17] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, test_fn::Nothing, julia_args::Cmd, test_args::Cmd, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool, kwargs::@Kwargs{io::Base.TTY})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:444
 [18] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.TTY, kwargs::@Kwargs{})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:159
 [19] test(pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:148
 [20] test(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::@Kwargs{})
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:174
 [21] test()
    @ Pkg.API ~/Software/julia-1.10.9/share/julia/stdlib/v1.10/Pkg/src/API.jl:165
 [22] top-level scope
     @ none:1
```

simone-silvestri commented:
Distributed tests should be run under MPI, and it seems that you are running distributed tests outside of the MPI executable (i.e with `julia --project` instead of `mpiexec -np 4 julia --project`).

francispoulin commented:
Also, it seems that the flux regressions are broke.  I don't know if that's causing a problem here.

```
Test Summary:     | Broken  Total     Time
Fluxes regression |      8      8  3m55.2s
WARNING: Method definition test_setting_from_metadata(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:32 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_setting_from_metadata(Any, Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:32 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_setting_from_metadata(Any, Any, Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:32 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_timestepping_with_dataset(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:54 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_ocean_metadata_utilities(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:81 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_dataset_restoring(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:115 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_timestepping_with_dataset_restoring(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:160 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_cycling_dataset_restoring(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:182 overwritten on the same line (check for duplicate calls to `include`).
WARNING: Method definition test_inpainting_algorithm(Any, Any, Any, Any) in module Main at /u/fpoulin/software/ClimaOcean.jl/test/runtests_setup.jl:234 overwritten on the same line (check for duplicate calls to `include`).
```

francispoulin committed:
Update make.jl

Removing the example to make sure everything else works file.

francispoulin committed:
Update make.jl

Reverting back

francispoulin committed:
Update make.jl

Resort to what we have in main, exactly. Surely this must work?

francispoulin committed:
Update runtests.jl

I want to see if everything passes without this one test.  I will return this afterwards.

francispoulin committed:
Update runtests.jl

Since many tests failed I am trying what we had before.  Very strange!

glwagner commented on docs/make.jl:
```diff
@@ -18,15 +18,19 @@ const EXAMPLES_DIR = joinpath(@__DIR__, "..", "examples")
 const OUTPUT_DIR   = joinpath(@__DIR__, "src/literated")
 
 to_be_literated = [
+    #"panantarctic_regional_simulation.jl",
```
```suggestion
    "panantarctic_regional_simulation.jl",
```
the example is commented out

glwagner commented on docs/make.jl:
```diff
@@ -18,15 +18,19 @@ const EXAMPLES_DIR = joinpath(@__DIR__, "..", "examples")
     filepath = joinpath(EXAMPLES_DIR, file)
     withenv("JULIA_DEBUG" => "Literate") do
         Literate.markdown(filepath, OUTPUT_DIR; flavor = Literate.DocumenterFlavor(), execute = true)
+        #run(`julia --project=docs -e "using Literate;
+        #    Literate.markdown(\"$filepath\", \"$outputpath\";
+        #                       flavor=Literate.DocumenterFlavor(), execute=true)"`)
```
```suggestion
        run(`julia --project=docs -e "using Literate;
            Literate.markdown(\"$filepath\", \"$outputpath\";
                               flavor=Literate.DocumenterFlavor(), execute=true)"`)
```

this step is necessary

glwagner commented on docs/make.jl:
```diff
@@ -47,6 +51,7 @@ pages = [
         "Single-column ocean simulation" => "literated/single_column_os_papa_simulation.md",
         "One-degree ocean--sea ice simulation" => "literated/one_degree_simulation.md",
         "Near-global ocean simulation" => "literated/near_global_ocean_simulation.md",
+        #"Panantarctic regional simulation" => "literated/panantarctic_regional_simulation.md",
```
```suggestion
        "Panantarctic regional simulation" => "literated/panantarctic_regional_simulation.md",
```

uncomment this to include the example in the docs

francispoulin replied:
Thanks!  Maybe I'll keep this commented for the moment, to make sure that everything passes without this example.  If yes, then I'll add the exampe back in.

glwagner commented:
it looks like there were some vestigial commented out things in `make.jl` which may have lead to failures.

francispoulin committed:
Update docs/make.jl

Co-authored-by: Gregory L. Wagner <wagner.greg@gmail.com>

francispoulin committed:
Update docs/make.jl

Co-authored-by: Gregory L. Wagner <wagner.greg@gmail.com>

francispoulin commented:
Thanks Greg!  Sorry, I thought I had returned things to normal, but clearly not the case.

I accepted two of your three suggestions, and want to see if all tests pass without this example being added to the docs.  If yes, then I'll add the example in to see what changes.

glwagner committed:
Update docs/make.jl

glwagner committed:
Update docs/make.jl

francispoulin commented:
Interestingly, when the new example is competely removed (thank you again @glwagner), the docs still fail.  Sadly, I still can't see the errors.

I am trying to build the docs on a server and I will let you know if I come across errors of any kind.

francispoulin commented:
When I tried running the `make.jl` script on another server to make the docs, I don't get an error.  I haven't looked at any of them but it seems to have created all of them, see below, including the new example without a single error.

Question: why does it fail on these two checks on github?

```
shell> ls -lrt src/literated/
total 90056
-rw-rw-r-- 1 fpoulin fpoulin   276168 Sep 15 16:06 panantarctic_bathymetry.png
-rw-rw-r-- 1 fpoulin fpoulin  1590517 Sep 15 16:54 acc_snapshot.png
-rw-rw-r-- 1 fpoulin fpoulin 18169005 Sep 15 16:54 panantarctic_regional_surface.mp4
-rw-rw-r-- 1 fpoulin fpoulin    18135 Sep 15 16:54 panantarctic_regional_simulation.md
-rw-rw-r-- 1 fpoulin fpoulin  1747402 Sep 15 17:01 single_column_profiles.mp4
-rw-rw-r-- 1 fpoulin fpoulin   596778 Sep 15 17:01 single_column_os_papa_simulation.md
-rw-rw-r-- 1 fpoulin fpoulin   598273 Sep 15 18:05 global_snapshot.png
-rw-rw-r-- 1 fpoulin fpoulin 50084758 Sep 15 18:06 one_degree_global_ocean_surface.mp4
-rw-rw-r-- 1 fpoulin fpoulin   696041 Sep 15 18:06 one_degree_simulation.md
-rw-rw-r-- 1 fpoulin fpoulin   505603 Sep 15 18:07 bathymetry.png
-rw-rw-r-- 1 fpoulin fpoulin  2891474 Sep 15 19:13 snapshot.png
-rw-rw-r-- 1 fpoulin fpoulin 15000775 Sep 15 19:13 near_global_ocean_surface.mp4
-rw-rw-r-- 1 fpoulin fpoulin    17508 Sep 15 19:13 near_global_ocean_simulation.md
```

glwagner commented:
It's an out-of-memory error on GPU

To display the buildkite publically we probably need to fiddle with buildkite settings

francispoulin committed:
Update panantarctic_regional_simulation.jl

Let's try a coarser grid to see if this allows us to avoid the memory errors.

francispoulin commented:
Ah, good to know!

I'm trying a coarser grid, 720x120x40, to see if this avoids the error.

francispoulin commented:
Funny that when I reduce the resolution, and don't do anything else differently, now we have more errors.  See below.

I'll return the parameters since this clearly did not help.

```
��� Warning: Opening file with JLD2.MmapIO failed, falling back to IOStream
��� @ JLD2 /central/scratch/esm/slurm-buildkite/climaocean-ci/4801/depot/default/packages/JLD2/WDhXU/src/JLD2.jl:162
Field utilities: Error During Test at /central/scratch/esm/slurm-buildkite/climaocean-ci/4801/climaocean-ci/test/test_ecco2_daily.jl:51
  Got exception outside of a @test
  EOFError: read end of file
```

francispoulin committed:
Update panantarctic_regional_simulation.jl

Returning to the desired resolution since more errors happened with a coarser grid.  Hmm...

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

francispoulin committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
ExponentialCoordinate -> ExponentialDiscretization

navidcy committed:
no need to spinup with RK3

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

francispoulin committed:
saving data to different folder based on arch

francispoulin committed:
Merge branch 'main' into fjp/acc_regional_model

francispoulin commented:
Any idea why we get this error when running distributed tests?

```srun: error: hpc-34-32: task 0: Exited with exit code 1
🚨 Error: The command exited with status 1
```

Also, can anyone see the error from the docs?  I get a login page but can't login.  (@cmbengue ?)

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model

navidcy committed:
Merge branch 'main' into fjp/acc_regional_model
